File: backend/app/utils/parser.py
"""
Amazon review HTML parser
Extracts review data from Amazon review pages
"""
from bs4 import BeautifulSoup
from typing import List, Dict, Optional
from datetime import datetime
import re
import logging

logger = logging.getLogger(__name__)


class AmazonReviewParser:
    """Parser for Amazon customer review pages"""

    def __init__(self, marketplace: str = "us"):
        self.marketplace = marketplace
        self.domain_map = {
            "us": "amazon.com",
            "de": "amazon.de",
            "uk": "amazon.co.uk",
            "fr": "amazon.fr",
            "it": "amazon.it",
            "es": "amazon.es",
            "ca": "amazon.ca",
            "jp": "amazon.co.jp",
        }

    def parse_html(self, html_content: str, asin: str) -> List[Dict]:
        """
        Parse Amazon review page HTML and extract review data

        Args:
            html_content: HTML content of the review page
            asin: Product ASIN

        Returns:
            List of review dictionaries
        """
        soup = BeautifulSoup(html_content, "lxml")
        reviews = []

        # Find all review containers
        review_divs = soup.find_all("div", {"data-hook": "review"})

        if not review_divs:
            logger.warning(f"No reviews found in HTML for ASIN: {asin}")
            return reviews

        for review_div in review_divs:
            try:
                review_data = self._parse_review_div(review_div, asin)
                if review_data:
                    reviews.append(review_data)
            except Exception as e:
                logger.error(f"Error parsing review: {e}")
                continue

        logger.info(f"Parsed {len(reviews)} reviews for ASIN: {asin}")
        return reviews

    def _parse_review_div(self, review_div, asin: str) -> Optional[Dict]:
        """Parse individual review div element"""

        # Review ID
        review_id = review_div.get("id", "").replace("customer_review-", "")
        if not review_id:
            return None

        # Rating (extract from star rating class)
        rating_elem = review_div.find("i", {"data-hook": "review-star-rating"})
        rating = None
        if rating_elem:
            rating_text = rating_elem.get_text(strip=True)
            match = re.search(r"(\d+\.?\d*)", rating_text)
            if match:
                rating = int(float(match.group(1)))

        # Title
        title_elem = review_div.find("a", {"data-hook": "review-title"})
        title = None
        if title_elem:
            title = title_elem.get_text(strip=True)
            # Remove "5.0 out of 5 stars" prefix if present
            title = re.sub(r"^\d+\.?\d*\s+out of \d+ stars\s*", "", title)

        # Review text
        text_elem = review_div.find("span", {"data-hook": "review-body"})
        text = text_elem.get_text(strip=True) if text_elem else None

        # Author name
        author_elem = review_div.find("span", {"class": "a-profile-name"})
        author_name = author_elem.get_text(strip=True) if author_elem else None

        # Review date
        date_elem = review_div.find("span", {"data-hook": "review-date"})
        review_date = None
        if date_elem:
            date_text = date_elem.get_text(strip=True)
            review_date = self._parse_date(date_text)

        # Verified purchase
        verified_elem = review_div.find("span", {"data-hook": "avp-badge"})
        verified_purchase = verified_elem is not None

        # Helpful votes
        helpful_elem = review_div.find("span", {"data-hook": "helpful-vote-statement"})
        helpful_votes = 0
        if helpful_elem:
            helpful_text = helpful_elem.get_text(strip=True)
            match = re.search(r"(\d+)", helpful_text)
            if match:
                helpful_votes = int(match.group(1))

        return {
            "review_id": review_id,
            "asin": asin,
            "marketplace": self.marketplace,
            "rating": rating,
            "title": title,
            "text": text,
            "author_name": author_name,
            "review_date": review_date,
            "verified_purchase": verified_purchase,
            "helpful_votes": helpful_votes,
        }

    def _parse_date(self, date_text: str) -> Optional[datetime]:
        """Parse review date from text"""
        try:
            # Remove "Reviewed in [Country] on " prefix
            date_text = re.sub(r"^.*?\bon\s+", "", date_text)

            # Common date formats
            date_formats = [
                "%B %d, %Y",  # January 15, 2024
                "%d %B %Y",  # 15 January 2024 (EU format)
                "%Y年%m月%d日",  # Japanese format
            ]

            for fmt in date_formats:
                try:
                    return datetime.strptime(date_text.strip(), fmt)
                except ValueError:
                    continue

            logger.warning(f"Could not parse date: {date_text}")
            return None

        except Exception as e:
            logger.error(f"Error parsing date '{date_text}': {e}")
            return None

    def extract_asin_from_url(self, url: str) -> Optional[str]:
        """Extract ASIN from Amazon product URL"""
        # Pattern: /dp/ASIN or /product/ASIN
        match = re.search(r"/(dp|product)/([A-Z0-9]{10})", url)
        if match:
            return match.group(2)

        # Check if it's already an ASIN
        if len(url) == 10 and url.isalnum():
            return url.upper()

        return None

    def extract_product_info(self, html_content: str, asin: str) -> Dict:
        """Extract product metadata from review page"""
        soup = BeautifulSoup(html_content, "lxml")

        # Product title
        title_elem = soup.find("h1", {"data-hook": "product-link"})
        title = title_elem.get_text(strip=True) if title_elem else None

        # Average rating
        avg_rating_elem = soup.find("span", {"data-hook": "rating-out-of-text"})
        avg_rating = None
        if avg_rating_elem:
            match = re.search(r"(\d+\.?\d*)", avg_rating_elem.get_text())
            if match:
                avg_rating = float(match.group(1))

        # Total reviews
        total_reviews_elem = soup.find("div", {"data-hook": "total-review-count"})
        total_reviews = None
        if total_reviews_elem:
            match = re.search(r"([\d,]+)", total_reviews_elem.get_text())
            if match:
                total_reviews = int(match.group(1).replace(",", ""))

        return {
            "asin": asin,
            "title": title,
            "average_rating": avg_rating,
            "total_reviews": total_reviews,
        }

File: chrome-extension/content/parser.js
/**
 * Amazon Review Parser
 * Parses Amazon review pages and extracts review data
 */

const AmazonReviewParser = {
  /**
   * Detect marketplace from current URL
   */
  detectMarketplace() {
    const hostname = window.location.hostname;
    const marketplaceMap = {
      'amazon.com': 'us',
      'amazon.de': 'de',
      'amazon.co.uk': 'uk',
      'amazon.fr': 'fr',
      'amazon.it': 'it',
      'amazon.es': 'es',
      'amazon.ca': 'ca',
      'amazon.co.jp': 'jp'
    };

    for (const [domain, code] of Object.entries(marketplaceMap)) {
      if (hostname.includes(domain)) {
        return code;
      }
    }
    return 'us'; // default
  },

  /**
   * Extract ASIN from URL
   */
  extractASIN() {
    const url = window.location.pathname;
    const match = url.match(/\/product-reviews\/([A-Z0-9]{10})/);
    return match ? match[1] : null;
  },

  /**
   * Extract product information from the page
   */
  extractProductInfo() {
    const asin = this.extractASIN();

    // Try multiple selectors for product title
    const titleSelectors = [
      'a[data-hook="product-link"]',
      '.a-link-normal[data-hook="product-link"]',
      'h1.a-size-large.a-text-ellipsis'
    ];

    let title = null;
    for (const selector of titleSelectors) {
      const elem = document.querySelector(selector);
      if (elem) {
        title = elem.textContent.trim();
        break;
      }
    }

    // Extract average rating
    const avgRatingElem = document.querySelector('[data-hook="rating-out-of-text"]');
    const avgRating = avgRatingElem ? parseFloat(avgRatingElem.textContent.match(/[\d.]+/)[0]) : null;

    // Extract total review count
    const totalReviewsElem = document.querySelector('[data-hook="total-review-count"]');
    const totalReviews = totalReviewsElem ? parseInt(totalReviewsElem.textContent.replace(/\D/g, '')) : null;

    return {
      asin,
      title,
      average_rating: avgRating,
      total_reviews: totalReviews,
      marketplace: this.detectMarketplace(),
      url: window.location.href
    };
  },

  /**
   * Parse a single review element
   */
  parseReview(reviewElement) {
    try {
      // Review ID
      const reviewId = reviewElement.getAttribute('data-review-id') ||
                      reviewElement.id ||
                      `review_${Date.now()}_${Math.random().toString(36).substr(2, 9)}`;

      // Rating
      const ratingElem = reviewElement.querySelector('i[data-hook="review-star-rating"], i[data-hook="cmps-review-star-rating"]');
      let rating = null;
      if (ratingElem) {
        const ratingText = ratingElem.textContent || ratingElem.getAttribute('class');
        const ratingMatch = ratingText.match(/(\d+(?:\.\d+)?)/);
        rating = ratingMatch ? parseFloat(ratingMatch[1]) : null;
      }

      // Title
      const titleElem = reviewElement.querySelector('[data-hook="review-title"]');
      const title = titleElem ? titleElem.textContent.trim().replace(/^[\d.]+\s*out of 5 stars\s*/i, '') : null;

      // Review text
      const textElem = reviewElement.querySelector('[data-hook="review-body"]');
      const text = textElem ? textElem.textContent.trim() : null;

      // Author
      const authorElem = reviewElement.querySelector('.a-profile-name');
      const authorName = authorElem ? authorElem.textContent.trim() : null;

      // Date
      const dateElem = reviewElement.querySelector('[data-hook="review-date"]');
      let reviewDate = null;
      if (dateElem) {
        const dateText = dateElem.textContent.trim();
        // Extract date from text like "Reviewed in Germany on December 15, 2023"
        const dateMatch = dateText.match(/on\s+(.+)$/i);
        if (dateMatch) {
          reviewDate = new Date(dateMatch[1]).toISOString().split('T')[0];
        }
      }

      // Verified purchase
      const verifiedElem = reviewElement.querySelector('[data-hook="avp-badge"]');
      const verifiedPurchase = verifiedElem !== null;

      // Helpful votes
      const helpfulElem = reviewElement.querySelector('[data-hook="helpful-vote-statement"]');
      let helpfulVotes = 0;
      if (helpfulElem) {
        const helpfulText = helpfulElem.textContent;
        const helpfulMatch = helpfulText.match(/(\d+)/);
        helpfulVotes = helpfulMatch ? parseInt(helpfulMatch[1]) : 0;
      }

      return {
        review_id: reviewId,
        rating,
        title,
        text,
        author_name: authorName,
        review_date: reviewDate,
        verified_purchase: verifiedPurchase,
        helpful_votes: helpfulVotes,
        marketplace: this.detectMarketplace(),
        asin: this.extractASIN()
      };
    } catch (error) {
      console.error('Error parsing review:', error);
      return null;
    }
  },

  /**
   * Parse all reviews on the current page
   */
  parseAllReviews() {
    const reviewSelectors = [
      '[data-hook="review"]',
      '.review',
      '[id^="customer_review"]'
    ];

    let reviews = [];
    for (const selector of reviewSelectors) {
      const reviewElements = document.querySelectorAll(selector);
      if (reviewElements.length > 0) {
        reviewElements.forEach(elem => {
          const review = this.parseReview(elem);
          if (review && review.rating) {
            reviews.push(review);
          }
        });
        break; // Found reviews, no need to try other selectors
      }
    }

    return reviews;
  },

  /**
   * Get pagination info
   */
  getPaginationInfo() {
    const paginationBar = document.querySelector('.a-pagination');
    if (!paginationBar) {
      return { currentPage: 1, totalPages: 1, hasNext: false };
    }

    // Current page
    const currentPageElem = paginationBar.querySelector('.a-selected');
    const currentPage = currentPageElem ? parseInt(currentPageElem.textContent) : 1;

    // Next page link
    const nextPageLink = paginationBar.querySelector('.a-last:not(.a-disabled) a');
    const hasNext = nextPageLink !== null;

    // Try to find total pages
    const pageLinks = paginationBar.querySelectorAll('.a-normal');
    const totalPages = pageLinks.length > 0 ?
      Math.max(...Array.from(pageLinks).map(link => parseInt(link.textContent))) :
      currentPage;

    return {
      currentPage,
      totalPages,
      hasNext,
      nextPageUrl: hasNext ? nextPageLink.href : null
    };
  },

  /**
   * Get complete page data
   */
  getPageData() {
    return {
      product: this.extractProductInfo(),
      reviews: this.parseAllReviews(),
      pagination: this.getPaginationInfo(),
      scrapedAt: new Date().toISOString()
    };
  }
};

// Make parser available globally for content script
window.AmazonReviewParser = AmazonReviewParser;
